xdgopenproxy/xdgopenproxy.go

```diff
@@ -27,37 +27,95 @@ import (
 	"github.com/godbus/dbus"
 )
 
-func Run(urlOrFile string) error {
-	bus, err := dbus.SessionBus()
+type bus interface {
+	Object(name string, objectPath dbus.ObjectPath) dbus.BusObject
+	Close() error
+}
+
+var sessionBus = func() (bus, error) { return dbus.SessionBus() }
+
+type desktopLauncher interface {
+	openFile(path string) error
+	openURL(url string) error
+}
+
+type portalLauncher struct {
+	service dbus.BusObject
+}
+
+func (p *portalLauncher) openFile(filename string) error {
+	fd, err := syscall.Open(filename, syscall.O_RDONLY, 0)
 	if err != nil {
 		return err
 	}
-	defer bus.Close()
-	launcher := bus.Object("io.snapcraft.Launcher", "/io/snapcraft/Launcher")
-	return launch(launcher, urlOrFile)
+	defer syscall.Close(fd)
+
+	return p.service.Call("org.freedesktop.portal.OpenURI.OpenFile", 0, "", dbus.UnixFD(fd),
```

Comment:
If the file in question is a directory we should be calling `OpenDirectory` instead, passing an `O_PATH` file descriptor.

Reply:
I don't see `OpenDirectory()` mentioned here https://flatpak.github.io/xdg-desktop-portal/portal-docs.html#gdbus-org.freedesktop.portal.OpenURI Passing directory to `OpenFile()` seems to work fine triggers `nautilus` window to be opened for that path. Maybe the method was in some previous iteration of the portal spec?

Reply:
I guess it is a new addition:

https://github.com/flatpak/xdg-desktop-portal/blob/master/data/org.freedesktop.portal.OpenURI.xml#L134

Maybe it would make sense to try `OpenDirectory` and fall back to `OpenFile` if we get NoSuchMethod.  I believe the reason for the new method is that it is impossible to open a directory FD with write access, so we can't use FD flags to prove that the caller has write access.  So it is split out into the separate call with slightly different checking.